app/controllers/purchases_controller.rb

```diff
@@ -9,94 +9,10 @@ def create
 
       if user.valid?
-        order = Order.new(
-          user: user,
-          first_name: user.first_name,
-          last_name: user.last_name,
-          address_1: address_params[:address_1],
-          address_2: address_params[:address_2],
-          city: address_params[:city],
-          state: address_params[:state],
-          country: address_params[:country],
-          zip: address_params[:zip],
-        )
-
-        cart.items.each do |item|
-          item.quantity.times do
-            order.items << OrderLineItem.new(
-              order: order,
-              sale: item.sale,
-              unit_price_cents: item.sale.unit_price_cents,
-              shipping_costs_cents: shipping_costs,
-              paid_price_cents: item.sale.unit_price_cents + shipping_costs
-            )
-          end
-        end
-
-        order.save
-
-        if order.valid?
-          return render json: { status: :success, order: { id: order.id } }, status: :ok
-        else
-          return render json: { errors: order.errors.map(&:full_message).map { |message|
-                                          { message: message }
-                                        } }, status: :unprocessable_entity
-        end
-      else
-        return render json: { errors: user.errors.map(&:full_message).map { |message|
-                                        { message: message }
-                                      } }, status: :unprocessable_entity
-      end
-    elsif purchase_params[:gateway] == 'stripe'
-      cart_id = purchase_params[:cart_id]
-
-      cart = Cart.find_by(id: cart_id)
-
-      unless cart
-        return render json: { errors: [{ message: 'Cart not found!' }] }, status: :unprocessable_entity
-      end
-
-      user = if cart.user.nil?
-               user_params = purchase_params[:user] ? purchase_params[:user] : {}
-               User.create(**user_params.merge(guest: true))
-             else
-               cart.user
-             end
-
-      if user.valid?
-        order = Order.new(
-          user: user,
-          first_name: user.first_name,
-          last_name: user.last_name,
-          address_1: address_params[:address_1],
-          address_2: address_params[:address_2],
-          city: address_params[:city],
-          state: address_params[:state],
-          country: address_params[:country],
-          zip: address_params[:zip],
-        )
-
-        cart.items.each do |item|
-          item.quantity.times do
-            order.items << OrderLineItem.new(
-              order: order,
-              sale: item.sale,
-              unit_price_cents: item.sale.unit_price_cents,
-              shipping_costs_cents: shipping_costs,
-              paid_price_cents: item.sale.unit_price_cents + shipping_costs
-            )
-          end
-        end
-
-        order.save
+        order = OrderCreator.call(user, address_params, cart)
 
         if order.valid?
```

Comment:
essa verificação se order é valida não irá funcionar aqui.
Teoricamente tu ja criou a order la no OrderCreate, se a order for inválida vai estourar uma exception lá.
O que tu pode verificar aqui é capturar a exception e renderizar o erro.
Uma dúvida: Se a order for inválida, o usuário continuara precisando ser criado?

Reply:
Se a criação do usuário depender de uma order valida, eu sugiro criar um service para tratar isso, e dentro do service tu chama a OrderCreate e UserCreate

Reply:
Deixar o controller o mais limpo possível de lógica de negócios